test/jdk/java/net/httpclient/CancelStreamedBodyTest.java

```diff
@@ -0,0 +1,432 @@
+        if (!share) return makeNewClient(share);
+        HttpClient shared = sharedClient;
+        if (shared != null) return shared;
+        synchronized (this) {
+            shared = sharedClient;
+            if (shared == null) {
+                shared = sharedClient = makeNewClient(share);
+            }
+            return shared;
+        }
+    }
+
+    final static String BODY = "Some string |\n that ?\n can |\n be split ?\n several |\n ways.";
+
+
+    @Test(dataProvider = "urls")
+    public void testAsLines(String uri, boolean sameClient)
+            throws Exception {
+        checkSkip();
+        HttpClient client = null;
+        uri = uri + "/testAsLines";
+        out.printf("%n%s testAsLines(%s, %b)%n", now(), uri, sameClient);
+        for (int i=0; i< ITERATION_COUNT; i++) {
+            if (!sameClient || client == null)
+                client = newHttpClient(sameClient);
+
+            HttpRequest req = HttpRequest.newBuilder(URI.create(uri))
+                    .GET()
+                    .build();
+            Exception failed = null;
+            List<String> lines = null;
+            for (int j = 0; j < 2; j++) {
+                try (Stream<String> body = client.send(req, BodyHandlers.ofLines()).body()) {
+                    lines = body.limit(j).toList();
+                    assertEquals(lines, BODY.replaceAll("\\||\\?", "")
+                            .lines().limit(j).toList());
+                }
+                var error = TRACKER.check(500,
+                        (t) -> t.getOutstandingOperations() > 0 || t.getOutstandingSubscribers() > 0,
+                        "subscribers for testAsLines(%s)\n\t step [%s,%s]".formatted(req.uri(), i,j),
+                        false);
+                Reference.reachabilityFence(client);
+                if (error != null) throw error;
+            }
+            if (sameClient) continue;
+            client = null;
+            System.gc();
+            var error = TRACKER.check(500);
+            if (error != null) throw error;
+        }
+    }
+
+    @Test(dataProvider = "urls")
+    public void testInputStream(String uri, boolean sameClient)
+            throws Exception {
+        checkSkip();
+        HttpClient client = null;
+        uri = uri + "/testInputStream";
+        out.printf("%n%s testInputStream(%s, %b)%n", now(), uri, sameClient);
+        for (int i=0; i< ITERATION_COUNT; i++) {
+            if (!sameClient || client == null)
+                client = newHttpClient(sameClient);
+
+            HttpRequest req = HttpRequest.newBuilder(URI.create(uri))
+                    .GET()
+                    .build();
+            int read = -1;
+            for (int j = 0; j < 2; j++) {
+                try (InputStream is = client.send(req, BodyHandlers.ofInputStream()).body()) {
+                    for (int k = 0; k < j; k++) {
+                        read = is.read();
+                        assertEquals(read, BODY.charAt(k));
```

Comment:
Hello Daniel, I don't follow this loop. The number of times it is iterating seems to be dependent on a statically decided value of `j` (which can't exceed 2). Each iteration it reads 1 byte and compares that byte with the pre-defined response content. Shouldn't this loop instead be doing `k < BODY.length();`?

Reply:
Now that I think with a fresher mind, I see what this loop does - it intentionally doesn't read the entire response content and instead closes the stream before the whole response is read, to trigger the original issue.